feat: add documentation regarding the placeholder mechanism (#241)
AmadeusITGroup/otter: docs/components/PLACEHOLDERS.md

The component has only 1 input and support a *content value*.

- The **id** input is required and will determine the unique identification for a specific placeholder component.
- The *content value* (or child nodes), if provided, will be used by the placeholder component as default display when no template has been found.

## How it works

Based on the **id** provided to the placeholder component, it will register itself to the event coming from **placeholderTemplate** and will display the template corresponding to the its ID in the store.

> **note**: it is **strongly encouraged** to use the placeholder mechanism in concert with the [Rules Engine](../rules-engine/) following [this documentation](../rules-engine/how-to-use/placeholders.md).

## How to define a placeholder template

The placeholder template is defined in a JSON file following this JSON Schema [placeholder-template.schema.json](../../packages/@o3r/components/schemas/placeholder-template.schema.json).

Example:

```json
{
  "template": "<p>My fact : <%= myFact %></p>",
  "vars": {
    "myFact ": {
      "value": "myFact ",
      "type": "fact"
    }
  }
}
```

> Then the JSON File can be downloaded and dispatched to the store as following:

Reviewer: ```suggestion
Then the JSON File can be downloaded and dispatched to the store as follows:
```

```typescript
import { setPlaceholderTemplateEntityFromUrl } from '@o3r/components';

const JsonFileUrl = 'url-to-the-json-file';

store.dispatch(setPlaceholderTemplateEntityFromUrl({
  call: (await fetch(JsonFileUrl)).json(),
  id: 'my-placeholder-template',
  resolvedUrl: JsonFileUrl,
  url: JsonFileUrl
}));
```

Or the object can be directly dispatched via the `setPlaceholderTemplateEntity` action.

> ## How integrate a placeholder in the apps

Reviewer: ```suggestion
## How to integrate a placeholder in the apps
```

> The following step should be follow in the application:

Reviewer: ```suggestion
The following steps should be followed in the application:
```

Based on the **id** provided to the placeholder component, it will register itself to the event coming from **placeholderTemplate** and will display the template corresponding to its ID in the store.

> **note**: it is **strongly encouraged** to use the placeholder mechanism in concert with the [Rules Engine](../rules-engine/) following [this documentation](../rules-engine/how-to-use/placeholders.md).

## How to define a placeholder template

The placeholder template is defined in a JSON file following this JSON Schema [placeholder-template.schema.json](../../packages/@o3r/components/schemas/placeholder-template.schema.json).

Example:

```json
{
  "template": "<p>My fact : <%= myFact %></p>",
  "vars": {
    "myFact ": {
      "value": "myFact ",
      "type": "fact"
    }
  }
}
```

Then the JSON File can be downloaded and dispatched to the store as follows:

```typescript
import { setPlaceholderTemplateEntityFromUrl } from '@o3r/components';

const JsonFileUrl = 'url-to-the-json-file';

store.dispatch(setPlaceholderTemplateEntityFromUrl({
  call: (await fetch(JsonFileUrl)).json(),
  id: 'my-placeholder-template',
  resolvedUrl: JsonFileUrl,
  url: JsonFileUrl
}));
```

> Or the object can be directly dispatched via the `setPlaceholderTemplateEntity` action.

Reviewer 2: the internal mechanism explained here has changed with https://github.com/AmadeusITGroup/otter/commit/e88d6eb0d980ed34c09bca358399d04bf934e9f2

### Static localization

The first choice you have when you want to localize your template is the static localization.

> You need to create a localized template for each locale and provide the template Url with `[LANGUAGE]` (ex: `assets/placeholders/[LANGUAGE]/myPlaceholder.json`)

Reviewer: ```suggestion
You need to create a localized template for each locale and provide the template URL with `[LANGUAGE]` (ex: `assets/placeholders/[LANGUAGE]/myPlaceholder.json`)
```

> The rules engine service will handle the replacement of `[LANGUAGE]` for you, and when you change lang a new call will be performed to the new 'translated' Url.

Reviewer: ```suggestion
The rules engine service will handle the replacement of `[LANGUAGE]` for you, and when you change language a new call will be performed to the new 'translated' URL.
```

> > **Note**: the URL caching mechanism is based on the url NOT 'translated', meaning that if you change from en-GB to fr-FR then you decide to switch back all the calls will be done again.

Reviewer: ```suggestion
> **Note**: the URL caching mechanism is based on the url NOT 'translated', meaning that if you change from en-GB to fr-FR then you decide to switch back and all the calls will be done again.
```

> > This behavior is based on the fact that a real user rarely go back and force with the language update.

Reviewer: ```suggestion
> This behavior is based on the fact that a real user rarely goes back and forth with the language update.
```
